src/internal_math.rs

```diff
@@ -1 +1,217 @@
+use std::mem::swap;
 
+/// # Arguments
+/// * `m` `1 <= m`
+///
+/// # Returns
+/// x mod m
+/* const */
+fn safe_mod(mut x: i64, m: i64) -> i64 {
+    x %= m;
+    if x < 0 {
+        x += m;
+    }
+    x
+}
+
+/// Fast moduler by barrett reduction
+/// Reference: https://en.wikipedia.org/wiki/Barrett_reduction
+/// NOTE: reconsider after Ice Lake
+struct Barrett {
+    _m: u32,
+    im: u64,
+}
+
+impl Barrett {
+    /// # Arguments
+    /// * `m` `1 <= m`
```

Comment:
Well, for `v = v.wrapping_add(self._m);` below to work correctly, m must be <= 2^31, although this precondition is missing in the original ACL.
Could you document this?

FYI: this will cause no trouble because `Barrett::new` is called only by `math`, where `m` is received as an `i32` and then cast to `u32`.

Reply:
@koba-e964 OK, I'll do it.  Just for curious, could you tell me why should that condition hold?  I didn't understand why `wrapping_add` does not work correctly for `w` more than 2^31.

Reply:
The `wrapping_add` line's intention is correction of `v`; if `v` is *negative* (by negative I mean it's theoretically negative: if there were no overflows/underflows, `z - x * self._m` would be a negative integer), we want to correct this by adding `self._m` so that `v` is in range `[0, self._m)`.
For example, if `m` is 3 * 2^30 and `v` is *in theory* -2^31, `v` is 2^31 if represented as `u32`. There is no way to tell if `v` is theoretically -2^31 or 2^31 because these two values have the same representation in `u32` and both are possible.
If `2 * self._m` is <= 2^32 this situation is impossible.

Reply:
Thank you for your detailed explanation!  I now understood.